pkg/seedadmission/pods.go

```diff
@@ -0,0 +1,132 @@
+
+	extensionscontroller "github.com/gardener/gardener/extensions/pkg/controller"
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	gardenlogger "github.com/gardener/gardener/pkg/logger"
+
+	"github.com/sirupsen/logrus"
+	"gomodules.xyz/jsonpatch/v2"
+	admissionv1beta1 "k8s.io/api/admission/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// MutatePod mutates pod if mutation conditions are met
+func MutatePod(ctx context.Context, c client.Client, logger *logrus.Logger, request *admissionv1beta1.AdmissionRequest) ([]jsonpatch.JsonPatchOperation, error) {
+	// Ignore all resources other than our expected ones
+	switch request.Resource {
+	case
+		metav1.GroupVersionResource{Group: corev1.SchemeGroupVersion.Group, Version: corev1.SchemeGroupVersion.Version, Resource: "pods"}:
+	default:
+		return nil, nil
+	}
+
+	obj, err := getRequestObject(ctx, c, request)
+	if err != nil {
+		return nil, client.IgnoreNotFound(err)
+	}
+
+	entryLogger := gardenlogger.
+		NewFieldLogger(logger, "resource", fmt.Sprintf("%s/%s/%s", request.Kind.Group, request.Kind.Version, request.Kind.Kind)).
+		WithField("operation", request.Operation).
+		WithField("namespace", request.Namespace)
+
+	entryLogger.Info("Handling request")
+	patch, err := mutateShootControlPlanePodAnnotations(ctx, c, entryLogger, obj, request.Namespace)
+	if err != nil {
+		logger.Errorf("Error mutating pod %s/%s: %v", request.Name, request.Namespace, err)
+		return nil, err
+	}
+
+	return patch, nil
+}
+
+// mutateShootControlPlanePodAnnotations checks if the object mutation is allowed and return the needed patches .
+func mutateShootControlPlanePodAnnotations(ctx context.Context, c client.Client, logger *logrus.Entry, pod runtime.Object, namespace string) ([]jsonpatch.JsonPatchOperation, error) {
+	var patch []jsonpatch.JsonPatchOperation
+	// determine whether to perform mutation
+	isMutationRequired, err := mutationRequired(ctx, c, namespace)
+	if err != nil {
+		return nil, err
+	}
+
+	acc, err := meta.Accessor(pod)
+	if err != nil {
+		return nil, err
+	}
+
+	if !isMutationRequired {
+		logger.Debugf("Skipping mutation for %s/%s due to policy check", acc.GetNamespace(), acc.GetName())
+		return nil, nil
+	}
+
+	originalAnnotations := acc.GetAnnotations()
+	annotationsToAdd := map[string]string{
+		"fluentbit.io/exclude": "true",
+	}
+
+	for key, value := range annotationsToAdd {
+		if originalAnnotations == nil || originalAnnotations[key] == "" {
+			patch = append(patch, jsonpatch.JsonPatchOperation{
+				Operation: "add",
+				Path:      "/metadata/annotations",
+				Value: map[string]string{
+					key: value,
+				},
+			})
+		} else {
+			patch = append(patch, jsonpatch.JsonPatchOperation{
+				Operation: "replace",
+				Path:      "/metadata/annotations/" + key,
```

Comment:
```suggestion
				Path:      "/metadata/annotations/" + strings.ReplaceAll(key, "/", "~1"),
```

@vlvasilev wrt to https://github.com/gardener/gardener/pull/2320/files#r428491172:

`~1` is used to escape `/` in a JSON patch. Without it this `replace` operation will fail:

```
error: pods "cloud-controller-manager-b5ccb7948-48vps" could not be patched: Internal error occurred: Internal error occurred: jsonpatch replace operation does not apply: doc is missing path: /metadata/annotations/fluentbit.io/exclude
```

You're right, it's not needed for the above `add` operation, but during replace you have to escape `/`. I tested and verified it locally.

Reply:
Thanks for the explanation.